File: Data Structures & Algorithms/trapping-rain-water/submission-2.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        res = 0
        lh = [0]*n
        rh = [0]*n
        for i in range(len(height)):
            lh[i] = max(lh[i-1], height[i]) if i > 0 else height[i]
        for i in range(len(height)-1, -1, -1):
            rh[i] = max(rh[i+1], height[i]) if i < len(height)-1 else height[i]

        for i in range(len(height)):
            res += min(lh[i], rh[i])-height[i]

        return res

File: Data Structures & Algorithms/maximum-sum-circular-subarray/submission-2.rs
impl Solution {
    pub fn max_subarray_sum_circular(nums: Vec<i32>) -> i32 {
        let mut max = nums[0];
        let mut min = nums[0];
        let mut cur_min = 0;
        let mut cur_max = 0;
        let mut sum = 0;
        let n = nums.len();
        for i in 0..n {
            if cur_max < 0 {
                cur_max = 0;
            }
            if cur_min > 0 {
                cur_min = 0;
            }
            cur_min += nums[i];
            cur_max += nums[i];
            sum += nums[i];
            max = max.max(cur_max);
            min = min.min(cur_min);
        }

        if max < 0 {max} else {max.max(sum - min)}
    }
}
